File: Content/Code/hash-set-open-addressing/python/open_addressing_set_exercise.py
from enum import Enum, auto


class ProbingType(Enum):
    LINEAR = auto()
    QUADRATIC = auto()
    DOUBLE_HASHING = auto()


class OpenAddressingIntSet:
    class _SlotState(Enum):
        EMPTY = auto()
        OCCUPIED = auto()
        DELETED = auto()

    def __init__(self, capacity_exponent, probing_type):
        if capacity_exponent < 1 or capacity_exponent > 30:
            raise ValueError("capacity_exponent must be from 1 through 30")
        if not isinstance(probing_type, ProbingType):
            raise ValueError("probing_type must be a ProbingType")
        self.m = 1 << capacity_exponent
        self._keys = [0] * self.m
        self._states = [self._SlotState.EMPTY] * self.m
        self._probing_type = probing_type
        self._count = 0

    def size(self):
        # TODO: Return the number of distinct keys currently stored.
        return -1

    def is_empty(self):
        # TODO: Return whether the set contains no keys.
        return False

    def capacity(self):
        return self.m

    def contains(self, key):
        # TODO: Reject a negative key. Probe at most capacity() positions,
        # stopping successfully at key or unsuccessfully at an EMPTY slot.
        return False

    def insert(self, key):
        # TODO: Reject a negative key. Remember the first DELETED slot, but
        # continue until finding key, an EMPTY slot, or the probe limit.
        # Add the key and increment _count exactly once only when it is absent.
        return False

    def remove(self, key):
        # TODO: Reject a negative key. Replace a matching OCCUPIED slot with
        # DELETED, decrement _count exactly once, and return whether key existed.
        return False

    def clear(self):
        # TODO: Restore every slot to EMPTY and reset _count to zero.
        pass

    def _probe_index(self, key, i):
        if self._probing_type is ProbingType.LINEAR:
            return self._linear_probe_index(key, i)
        if self._probing_type is ProbingType.QUADRATIC:
            return self._quadratic_probe_index(key, i)
        if self._probing_type is ProbingType.DOUBLE_HASHING:
            return self._double_hash_probe_index(key, i)
        raise AssertionError("unknown probing type")

    def _linear_probe_index(self, key, i):
        # TODO: Compute probe i using linear probing.
        return 0

    def _quadratic_probe_index(self, key, i):
        # TODO: Compute probe i using triangular quadratic offsets.
        return 0

    def _double_hash_probe_index(self, key, i):
        # TODO: Compute probe i using _primary_hash and _secondary_hash.
        return 0

    def _primary_hash(self, key):
        # TODO: Compute h1(key).
        return 0

    def _secondary_hash(self, key):
        # TODO: Compute the odd double-hashing step h2(key).
        return 0

    @staticmethod
    def _check_key(key):
        if key < 0:
            raise ValueError("key must be nonnegative")


failures = 0


def check(condition, label):
    global failures
    if condition:
        print(f"pass: {label}")
    else:
        failures += 1
        print(f"FAIL: {label}")


def check_equal(actual, expected, label):
    check(actual == expected, f"{label} (expected {expected}, got {actual})")


def check_raises(action, label):
    global failures
    try:
        action()
        failures += 1
        print(f"FAIL: {label}")
    except ValueError:
        print(f"pass: {label}")


def test_strategy(probing_type):
    name = probing_type.name
    values = OpenAddressingIntSet(3, probing_type)
    check_equal(values.capacity(), 8, f"{name}: exponent 3 gives capacity 8")
    check(values.is_empty() and values.size() == 0, f"{name}: new set is empty")
    check(values.insert(1) and values.insert(9) and values.insert(17),
          f"{name}: insert colliding keys")
    check(values.contains(1) and values.contains(9) and values.contains(17),
          f"{name}: find colliding keys")
    check(not values.insert(17) and values.size() == 3, f"{name}: reject duplicate insertion")
    check(values.remove(9) and not values.contains(9), f"{name}: remove creates a tombstone")
    check(values.contains(17), f"{name}: lookup continues past a tombstone")
    check(values.insert(41) and values.contains(41), f"{name}: insertion can reuse a tombstone")
    check(not values.remove(99), f"{name}: absent removal changes nothing")

    wraparound = OpenAddressingIntSet(3, probing_type)
    check(wraparound.insert(7) and wraparound.insert(15) and wraparound.contains(15),
          f"{name}: probe sequence wraps around")

    full = OpenAddressingIntSet(3, probing_type)
    filled = all(full.insert(key) for key in range(full.capacity()))
    check(filled and not full.insert(8), f"{name}: insertion fails when no position is available")
    check(full.remove(0) and full.insert(8) and full.contains(8),
          f"{name}: insertion reuses the only available tombstone")

    values.clear()
    check(values.is_empty() and values.size() == 0 and not values.contains(1),
          f"{name}: clear resets the set")
    check_raises(lambda: values.contains(-1), f"{name}: reject negative lookup key")
    check_raises(lambda: values.insert(-1), f"{name}: reject negative insertion key")
    check_raises(lambda: values.remove(-1), f"{name}: reject negative removal key")


def test_tombstone_stress(probing_type):
    name = f"{probing_type.name} stress"
    values = OpenAddressingIntSet(7, probing_type)
    m = values.capacity()
    result = True

    # All 80 keys have home index 3, so every insertion must resolve collisions.
    for q in range(80):
        result = values.insert(3 + m * q) and result
    check(result and values.size() == 80, f"{name}: insert 80 colliding keys")

    result = True
    for q in range(80):
        result = values.contains(3 + m * q) and result
    check(result, f"{name}: find all initial keys")

    # Remove 27 keys. Each replacement has the same home index and, for
    # double hashing, the same step as the key it replaces.
    result = True
    for q in range(0, 80, 3):
        result = values.remove(3 + m * q) and result
    check(result and values.size() == 53, f"{name}: remove 27 keys")

    result = True
    for q in range(80):
        key = 3 + m * q
        found_as_expected = not values.contains(key) if q % 3 == 0 else values.contains(key)
        result = found_as_expected and result
    check(result, f"{name}: searches cross tombstones correctly")

    result = True
    for q in range(0, 80, 3):
        result = values.insert(3 + m * (q + 128)) and result
    check(result and values.size() == 80, f"{name}: replace all 27 removed keys")

    # Create and refill a second wave of 14 tombstones.
    result = True
    for q in range(0, 80, 6):
        result = values.remove(3 + m * (q + 128)) and result
    check(result and values.size() == 66, f"{name}: remove 14 replacement keys")

    result = True
    for q in range(0, 80, 6):
        result = values.insert(3 + m * (q + 256)) and result
    check(result and values.size() == 80, f"{name}: refill the second tombstone wave")

    result = True
    for q in range(80):
        if q % 6 == 0:
            result = not values.contains(3 + m * (q + 128)) and result
            result = values.contains(3 + m * (q + 256)) and result
        elif q % 3 == 0:
            result = values.contains(3 + m * (q + 128)) and result
        else:
            result = values.contains(3 + m * q) and result
    check(result, f"{name}: final membership is correct after 41 removals and replacements")


if __name__ == "__main__":
    check_raises(lambda: OpenAddressingIntSet(0, ProbingType.LINEAR),
                 "reject an invalid capacity exponent")
    check_raises(lambda: OpenAddressingIntSet(3, None), "reject an invalid probing type")
    for strategy in ProbingType:
        test_strategy(strategy)
        test_tombstone_stress(strategy)
    print("All tests passed." if failures == 0 else f"{failures} test(s) failed.")

File: Content/Code/hash-set-open-addressing/java/OpenAddressingSetExercise.java
import java.util.Arrays;

public class OpenAddressingSetExercise {
    enum ProbingType { LINEAR, QUADRATIC, DOUBLE_HASHING }

    static final class OpenAddressingIntSet {
        private enum SlotState { EMPTY, OCCUPIED, DELETED }

        private final int m;
        private final int[] keys;
        private final SlotState[] states;
        private final ProbingType probingType;
        private int count;

        OpenAddressingIntSet(int capacityExponent, ProbingType probingType) {
            if (capacityExponent < 1 || capacityExponent > 30) {
                throw new IllegalArgumentException("capacityExponent must be from 1 through 30");
            }
            if (probingType == null) {
                throw new IllegalArgumentException("probingType must not be null");
            }
            m = 1 << capacityExponent;
            keys = new int[m];
            states = new SlotState[m];
            Arrays.fill(states, SlotState.EMPTY);
            this.probingType = probingType;
        }

        int size() {
            // TODO: Return the number of distinct keys currently stored.
            return -1;
        }

        boolean isEmpty() {
            // TODO: Return whether the set contains no keys.
            return false;
        }

        int capacity() { return m; }

        boolean contains(int key) {
            // TODO: Reject a negative key. Probe at most capacity() positions,
            // stopping successfully at key or unsuccessfully at an EMPTY slot.
            return false;
        }

        boolean insert(int key) {
            // TODO: Reject a negative key. Remember the first DELETED slot, but
            // continue until finding key, an EMPTY slot, or the probe limit.
            // Add the key and increment count exactly once only when it is absent.
            return false;
        }

        boolean remove(int key) {
            // TODO: Reject a negative key. Replace a matching OCCUPIED slot with
            // DELETED, decrement count exactly once, and return whether key existed.
            return false;
        }

        void clear() {
            // TODO: Restore every slot to EMPTY and reset count to zero.
        }

        private int probeIndex(int key, int i) {
            switch (probingType) {
                case LINEAR:
                    return linearProbeIndex(key, i);
                case QUADRATIC:
                    return quadraticProbeIndex(key, i);
                case DOUBLE_HASHING:
                    return doubleHashProbeIndex(key, i);
                default:
                    throw new AssertionError("unknown probing type");
            }
        }

        private int linearProbeIndex(int key, int i) {
            // TODO: Compute probe i using linear probing.
            return 0;
        }

        private int quadraticProbeIndex(int key, int i) {
            // TODO: Compute probe i using triangular quadratic offsets.
            return 0;
        }

        private int doubleHashProbeIndex(int key, int i) {
            // TODO: Compute probe i using primaryHash and secondaryHash.
            return 0;
        }

        private int primaryHash(int key) {
            // TODO: Compute h1(key).
            return 0;
        }

        private int secondaryHash(int key) {
            // TODO: Compute the odd double-hashing step h2(key).
            return 0;
        }

        private static void checkKey(int key) {
            if (key < 0) throw new IllegalArgumentException("key must be nonnegative");
        }
    }

    private static int failures;

    private static void check(boolean condition, String label) {
        if (condition) System.out.println("pass: " + label);
        else { failures++; System.out.println("FAIL: " + label); }
    }

    private static void check(int actual, int expected, String label) {
        check(actual == expected, label + " (expected " + expected + ", got " + actual + ")");
    }

    private static void checkThrows(Runnable action, String label) {
        try { action.run(); failures++; System.out.println("FAIL: " + label); }
        catch (IllegalArgumentException expected) { System.out.println("pass: " + label); }
    }

    private static void testStrategy(ProbingType type) {
        String name = type.toString();
        OpenAddressingIntSet set = new OpenAddressingIntSet(3, type);
        check(set.capacity(), 8, name + ": exponent 3 gives capacity 8");
        check(set.isEmpty() && set.size() == 0, name + ": new set is empty");
        check(set.insert(1) && set.insert(9) && set.insert(17), name + ": insert colliding keys");
        check(set.contains(1) && set.contains(9) && set.contains(17), name + ": find colliding keys");
        check(!set.insert(17) && set.size() == 3, name + ": reject duplicate insertion");
        check(set.remove(9) && !set.contains(9), name + ": remove creates a tombstone");
        check(set.contains(17), name + ": lookup continues past a tombstone");
        check(set.insert(41) && set.contains(41), name + ": insertion can reuse a tombstone");
        check(!set.remove(99), name + ": absent removal changes nothing");

        OpenAddressingIntSet wraparound = new OpenAddressingIntSet(3, type);
        check(wraparound.insert(7) && wraparound.insert(15) && wraparound.contains(15),
                name + ": probe sequence wraps around");

        OpenAddressingIntSet full = new OpenAddressingIntSet(3, type);
        boolean filled = true;
        for (int key = 0; key < full.capacity(); key++) filled &= full.insert(key);
        check(filled && !full.insert(8), name + ": insertion fails when no position is available");
        check(full.remove(0) && full.insert(8) && full.contains(8),
                name + ": insertion reuses the only available tombstone");

        set.clear();
        check(set.isEmpty() && set.size() == 0 && !set.contains(1), name + ": clear resets the set");
        checkThrows(() -> set.contains(-1), name + ": reject negative lookup key");
        checkThrows(() -> set.insert(-1), name + ": reject negative insertion key");
        checkThrows(() -> set.remove(-1), name + ": reject negative removal key");
    }

    private static void testTombstoneStress(ProbingType type) {
        String name = type + " stress";
        OpenAddressingIntSet set = new OpenAddressingIntSet(7, type);
        int m = set.capacity();
        boolean result = true;

        // All 80 keys have home index 3, so every insertion must resolve collisions.
        for (int q = 0; q < 80; q++) result &= set.insert(3 + m * q);
        check(result && set.size() == 80, name + ": insert 80 colliding keys");

        result = true;
        for (int q = 0; q < 80; q++) result &= set.contains(3 + m * q);
        check(result, name + ": find all initial keys");

        // Remove 27 keys. Each replacement has the same home index and, for
        // double hashing, the same step as the key it replaces.
        result = true;
        for (int q = 0; q < 80; q += 3) result &= set.remove(3 + m * q);
        check(result && set.size() == 53, name + ": remove 27 keys");

        result = true;
        for (int q = 0; q < 80; q++) {
            int key = 3 + m * q;
            result &= q % 3 == 0 ? !set.contains(key) : set.contains(key);
        }
        check(result, name + ": searches cross tombstones correctly");

        result = true;
        for (int q = 0; q < 80; q += 3) result &= set.insert(3 + m * (q + 128));
        check(result && set.size() == 80, name + ": replace all 27 removed keys");

        // Create and refill a second wave of 14 tombstones.
        result = true;
        for (int q = 0; q < 80; q += 6) result &= set.remove(3 + m * (q + 128));
        check(result && set.size() == 66, name + ": remove 14 replacement keys");

        result = true;
        for (int q = 0; q < 80; q += 6) result &= set.insert(3 + m * (q + 256));
        check(result && set.size() == 80, name + ": refill the second tombstone wave");

        result = true;
        for (int q = 0; q < 80; q++) {
            if (q % 6 == 0) {
                result &= !set.contains(3 + m * (q + 128));
                result &= set.contains(3 + m * (q + 256));
            } else if (q % 3 == 0) {
                result &= set.contains(3 + m * (q + 128));
            } else {
                result &= set.contains(3 + m * q);
            }
        }
        check(result, name + ": final membership is correct after 41 removals and replacements");
    }

    public static void main(String[] args) {
        checkThrows(() -> new OpenAddressingIntSet(0, ProbingType.LINEAR),
                "reject an invalid capacity exponent");
        checkThrows(() -> new OpenAddressingIntSet(3, null), "reject a null probing type");
        for (ProbingType type : ProbingType.values()) {
            testStrategy(type);
            testTombstoneStress(type);
        }
        System.out.println(failures == 0 ? "All tests passed."
                : failures + " test(s) failed.");
    }
}
